src/crimson/os/seastore/onode_manager/staged-fltree/value.h

```diff
@@ -0,0 +1,270 @@
+}
+
+/**
+ * value_header_t
+ *
+ * The header structure in value layout.
+ *
+ * Value layout:
+ *
+ * # <- alloc size -> #
+ * # header | payload #
+ */
+struct value_header_t {
+  value_magic_t magic;
+  value_size_t payload_size;
+
+  value_size_t allocation_size() const {
+    return payload_size + sizeof(value_header_t);
+  }
+
+  const char* get_payload() const {
+    return reinterpret_cast<const char*>(this) + sizeof(value_header_t);
+  }
+
+  NodeExtentMutable get_payload_mutable(NodeExtentMutable& node) const {
+    return node.get_mutable_absolute(get_payload(), payload_size);
+  }
+
+  char* get_payload() {
+    return reinterpret_cast<char*>(this) + sizeof(value_header_t);
+  }
+
+  void initiate(NodeExtentMutable& mut, const value_config_t& config) {
+    value_header_t header{config.magic, config.payload_size};
+    mut.copy_in_absolute(this, header);
+    mut.set_absolute(get_payload(), 0, config.payload_size);
+  }
+
+  static value_size_t estimate_allocation_size(value_size_t payload_size) {
+    return payload_size + sizeof(value_header_t);
+  }
+} __attribute__((packed));
+inline std::ostream& operator<<(std::ostream& os, const value_header_t& header) {
+  return os << "Value(" << header.magic
+            << ", " << header.payload_size << "B)";
+}
+
+inline value_size_t value_config_t::allocation_size() const {
+  return value_header_t::estimate_allocation_size(payload_size);
+}
+
+/**
+ * ValueDeltaRecorder
+ *
+ * An abstracted class to handle user-defined value delta encode, decode and
+ * replay.
+ */
+class ValueDeltaRecorder {
+ public:
+  virtual ~ValueDeltaRecorder() = default;
+  ValueDeltaRecorder(const ValueDeltaRecorder&) = delete;
+  ValueDeltaRecorder(ValueDeltaRecorder&&) = delete;
+  ValueDeltaRecorder& operator=(const ValueDeltaRecorder&) = delete;
+  ValueDeltaRecorder& operator=(ValueDeltaRecorder&&) = delete;
+
+  /// Returns the value header magic for validation purpose.
+  virtual value_magic_t get_header_magic() const = 0;
+
+  /// Called by DeltaRecorderT to apply user-defined value delta.
+  virtual void apply_value_delta(ceph::bufferlist::const_iterator&,
+                                 NodeExtentMutable&,
+                                 laddr_t) = 0;
+
+ protected:
+  ValueDeltaRecorder(ceph::bufferlist& encoded) : encoded{encoded} {}
+
+  /// Get the delta buffer to encode user-defined value delta.
+  ceph::bufferlist& get_encoded(NodeExtentMutable&);
+
+ private:
+  ceph::bufferlist& encoded;
+};
+
+class tree_cursor_t;
+/**
+ * Value
+ *
+ * Value is a stateless view of the underlying value header and payload content
+ * stored in a tree leaf node, with the support to implement user-defined value
+ * deltas and to extend and trim the underlying payload data (not implemented
+ * yet).
+ *
+ * In the current implementation, we don't guarantee any alignment for value
+ * payload due to unaligned node layout and the according merge and split
+ * operations.
```

Comment:
Not a problem for this PR, but I think we'll need to support a user defined alignment restriction.  I hope ultimately for reads to entirely avoid copying or decoding the onode data.

Reply:
I think it is possible to re-arrange and align the node-internal indexing data structures to make the value payload to have the same alignment property (e.g. 8 bytes). But it is more complex to support flexible user-defined value alignment, as the node insert/split/... operations may need to respect the potential different alignment other than the inherent alignment defined by internal indexing data structures, and under different layout designs.

Reply:
Doesn't need to be flexible, we'll only ever use this for the onode structure.